code/modules/plumbing/plumbers/reaction_chamber.dm

```diff
@@ -147,13 +172,31 @@
 	QDEL_NULL(alkaline_beaker)
 	return ..()
 
-/obj/machinery/plumbing/reaction_chamber/chem/process(seconds_per_tick)
-	//add acidic/alkaine buffer if over/under limit
-	if(reagents.is_reacting && reagents.ph < alkaline_limit)
-		alkaline_beaker.reagents.trans_to(reagents, 1 * seconds_per_tick)
-	if(reagents.is_reacting && reagents.ph > acidic_limit)
-		acidic_beaker.reagents.trans_to(reagents, 1 * seconds_per_tick)
-	..()
+/obj/machinery/plumbing/reaction_chamber/chem/handle_reagents(seconds_per_tick)
+	while(TRUE)
+		if(machine_stat & NOPOWER)
+			return
+
+		//not too acidic nor too basic. work done
+		if(reagents.ph >= acidic_limit && reagents.ph <= alkaline_limit)
+			return
+		/**
+		 * figure out which buffer to transfer to restore balance
+		 * if solution is getting too basic(high ph) add some acid to lower it's value
+		 * else if solution is getting too acidic(low ph) add some base to increase it's value
+		 */
+		var/datum/reagents/buffer = reagents.ph > alkaline_limit ? acidic_beaker.reagents : alkaline_beaker.reagents
+		if(!buffer.total_volume)
+			return
+
+		//transfer buffer and handle reactions, not a proven math but looks logical
+		var/transfer_amount = FLOOR((reagents.ph > alkaline_limit ? (reagents.ph - alkaline_limit) : (acidic_limit - reagents.ph)) * seconds_per_tick, CHEMICAL_QUANTISATION_LEVEL)
+		if(transfer_amount <= 0 || !buffer.trans_to(reagents, transfer_amount))
+			return
+		reagents.handle_reactions()
+
+		//some power for accurate ph balancing
+		use_power((active_power_usage / 5) * seconds_per_tick)
 
```

Comment:
Can you restructure this to not use a `while(TRUE)`? 
You can definitely just do this with a regular while()

Reply:
I could move the ph range check to the while condition sure